### ISSUE TITLE
Blacklist admin as bucket name

### ISSUE THREAD

**Issue by s3rj1k** [minio/minio#9090]
Creating bucket named as admin and then uploading file to `v2/list-canned-policies` breaks routing logic for per bucket load balancer.

As per https://docs.min.io/docs/setup-nginx-proxy-with-minio.html `Proxy requests based on the bucket`

Adding option to blacklist some string patterns as bucket names should fix issue then someone tries to create object named as defined admin action.

Another solution would be to add ability to bind admin endpoints to different address.

**Comment by harshavardhana** [comment]
@s3rj1k I don't think we support bucket based routing in nginx.

The admin API exists at /minio/admin/v2 so I don't know why it would break, it looks like a setup issue on your end.

Can you share your nginx config?

**Comment by s3rj1k** [comment]
@harshavardhana Example named `Proxy requests based on the bucket` provided in https://docs.min.io/docs/setup-nginx-proxy-with-minio.html

I am just extending this example with bucked named `admin`.

So actually you do advertise support for bucket based routing ;).

Things will work wrong when some one would require to add access ACLs on per bucket basis.
No simple way for a router (`nginx`) to distinguish between access to `admin/v2/list-canned-policies` object and admin action.

Preferred solution to this would be separate IP for admin endpoint.

Alternative one could use virtual-host-style routing, but this approach a bit harder to implement.

**Comment by harshavardhana** [comment]
That looks incorrect in documentation @s3rj1k will address it. Also using /admin shouldn't conflict with Admin API because /admin is not at /admin but instead at /minio/admin

**Comment by harshavardhana** [comment]
Please share your nginx config as previously requested..

**Comment by s3rj1k** [comment]
```
worker_processes auto;
worker_cpu_affinity auto;

worker_rlimit_nofile 800000;

user nginx;
pid /var/run/nginx.pid;

error_log /var/log/nginx/errors.log;

events {
  worker_connections 10240;
  use epoll;
}

http {
  include /etc/nginx/mime.types;
  default_type application/octet-stream;

  server_tokens off;

  upstream s3t1 {
    ip_hash;

    server s1t1.local:9000;
    server s2t1.local:9000;
    server s3t1.local:9000;
    server s4t1.local:9000;
  }

  upstream s3t2 {
    ip_hash;

    server s1t2.local:9000;
    server s2t2.local:9000;
    server s3t2.local:9000;
    server s4t2.local:9000;
  }

  server {
    listen 80;
    server_name _;

    ignore_invalid_headers off;

    location / {
      client_max_body_size 0;

      proxy_http_version 1.1;
      proxy_buffering off;
      proxy_connect_timeout 300;

      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;
      proxy_set_header Host $http_host;
      proxy_set_header Connection "";

      proxy_pass http://s3t1;
    }

    location /admin/ {
      client_max_body_size 0;

      proxy_http_version 1.1;
      proxy_buffering off;
      proxy_connect_timeout 300;

      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;
      proxy_set_header Host $http_host;
      proxy_set_header Connection "";

      proxy_pass http://s3t2;
    }
  }
}

```

Bucket named admin would route to s3t2 and all admin commands will also route to s3t2.
That makes s3t1 unmanageable from outside network.

This is only one example how routing based on bucket can cause unexpected behavior.

I hope that you would not delete this approach from docs and mark it as deprecated.

This is perfectly valid solution to get simple federated setup of multiple clusters.

**Comment by harshavardhana** [comment]
>     location /admin/ {

The issue here is perhaps the extra `/` in `/admin/`  if you keep it `/admin` it shouldn't route to s3t2 @s3rj1k

**Comment by s3rj1k** [comment]
The idea is to route parts of buckets to different cluster based on some automatic nginx config generation where buckets can be created by end user meaning cluster administrator has no control on naming of buckets.

Some end-user client then can abuse naming of buckets and this can create DoS for cluster administrator.

**Comment by harshavardhana** [comment]
> Some end-user client then can abuse naming of buckets and this can create DoS for cluster administrator.

I don't think should happen there is something else happening, it cannot be routed to `/minio/admin`  from `/admin` bucket, let me get back to you on this.

**Comment by harshavardhana** [comment]
@s3rj1k I am able to successfully use "/" and "/admin" without issues

```
~ mc ls myminio-local1/admin/v2/
[2020-03-05 20:03:28 PST]    282B list-canned-policies

~ mc admin info myminio-local1
●  localhost:9001
   Uptime: 1 minute 
   Version: 2020-03-05T22:03:35Z
   Network: 1/1 OK 

80 MiB Used, 3 Buckets, 407 Objects
```

```
~ mc admin policy list myminio-local1
diagnostics         
readonly            
readwrite           
writeonly           
```

```
worker_processes auto;
worker_cpu_affinity auto;

worker_rlimit_nofile 800000;
pid /tmp/nginx.pid;

events {
  worker_connections 10240;
  use epoll;
}

http {
  include /etc/nginx/mime.types;
  default_type application/octet-stream;

  server_tokens off;

  server {
    listen 9001;
    server_name localhost;

    ignore_invalid_headers off;

    location / {
      client_max_body_size 0;

      proxy_http_version 1.1;
      proxy_buffering off;
      proxy_connect_timeout 300;

      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;
      proxy_set_header Host $http_host;
      proxy_set_header Connection "";

      proxy_pass http://localhost:9000;
    }

    location /admin {
      client_max_body_size 0;

      proxy_http_version 1.1;
      proxy_buffering off;
      proxy_connect_timeout 300;

      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;
      proxy_set_header Host $http_host;
      proxy_set_header Connection "";

      proxy_pass https://play.minio.io:9000;
    }
  }
}
```

Nginx started as `nginx -c /tmp/proxy.conf`

**Comment by s3rj1k** [comment]
@harshavardhana Hmm, how about getting to admin console of both clusters? 

I am testing with this https://nginx.viraptor.info/

Then bucket `administrator` is created, according to your config and above location tester it would be routed to `play.minio.io:9000`. I do not think that this behaviour is intended.

**Comment by harshavardhana** [comment]
> @harshavardhana Hmm, how about getting to admin console of both clusters?

what do you mean both? you can't do that because you don't have a prefix to route to another cluster. You cannot manage two clusters by single alias proxy that is not the intention. 

> Then bucket `administrator` is created, according to your config and above location tester it would be routed to `play.minio.io:9000`. I do not think that this behaviour is intended.

Why isn't that intended? I am showing that what you said is not possible. There is no incorrect routing here because of `/admin`  and there is no blacklisting needed.

**Comment by s3rj1k** [comment]
@harshavardhana I am not seeing any problem routing to different cluster admin when control plane can be optionally separated from data plain, sadly this is not the case currently with `minio`, this severely narrows use cases for `minio`.

When I add custom route for `/admin` I do not expect that any other bucket that has similar prefix will be routed to this location, thus I add trailing `/` to ensure this kind of undefined behaviour will never occur in first place.

Your solution to a problem is to abuse routing prefixes, that demands clever sorting of locations in `nginx`, longest must be first.

In my opinion this problem should be handled by `minio` itself and not be load-balancer software.

To fix this in `minio` there are 2 options:

1. Add an optional configuration option to allow blacklisting bucket names by regexp, or by some kind of other filter.
2. Add ability to bind HTTP server for control plane, `/admin/v2` endpoint, to address that differs from address for data plane.

**Comment by harshavardhana** [comment]
@s3rj1k I think there is some understanding mistake - you are trying to do things with MinIO which it is not suitable.

> * Add an optional configuration option to allow blacklisting bucket names by regexp, or by some kind of other filter.

This is not acceptable as it allows unnecessary configuration styles. I think this usecase is better not addressed.

> * Add ability to bind HTTP server for control plane, `/admin/v2` endpoint, to address that differs from address for data plane.

This is not going to be acceptable because we cannot foresee ourselves opening multiple ports and asking admins to allow new ports to be whitelisted in their firewall.

**Comment by s3rj1k** [comment]
@harshavardhana Optional config is not forcing other admins to open new ports, you can easily extend `minio` to add additional functions and still leave default config as it was before.

**Comment by s3rj1k** [comment]
@harshavardhana With that kind of attitude to issues people are just going to use other solutions.

**Comment by harshavardhana** [comment]
> @harshavardhana Optional config is not forcing other admins to open new ports, you can easily extend `minio` to add additional functions and still leave default config as it was before.

Adding new options is a rabbit hole we wish to avoid as much as possible in general, there are already way too many ways to run MinIO - adding one more is not going to make it any better. If you certainly told us lets remove an option or two we would have more ears :-). 

> @harshavardhana With that kind of attitude to issues people are just going to use other solutions.

I don't mean to offend, we are more than welcome for a discussion and I certainly showed you what is currently possible. But in the end, we would certainly address this if this was an important requirement, for now, these types of use cases are not really MinIO's priority.

MinIO already does painfully a lot of things in a single static binary, if it's not suited for your needs you should use other solutions.

**Comment by s3rj1k** [comment]
@harshavardhana No problem, lets remove everything related to federation clustering with etcd and coredns and redo this option by using more simpler methods ;)

**Comment by harshavardhana** [comment]
> @harshavardhana No problem, lets remove everything related to federation clustering with etcd and coredns and redo this option by using more simpler methods ;)

you would be surprised - that is, in fact, the plan @s3rj1k federation was a stop gap before we added bucket expansion, which we already do now for scale-out distributed server deployments.

Federation only serves a limited purpose now to provide a sort of DNS + etcd solution when you create buckets so that you don't have to manually manage DNS.  But you are free to not use it and have your DNS provided be populated manually for your setup.

**Comment by harshavardhana** [comment]
Feel free to close this issue if we are in agreement @s3rj1k

**Comment by s3rj1k** [comment]
@harshavardhana  So proper and only supported way of doing federation is in fact virtual-host-style routing(https://docs.min.io/docs/minio-server-configuration-guide.html), please confirm?

**Comment by harshavardhana** [comment]
> @harshavardhana So proper and only supported way of doing federation is in fact virtual-host-style routing(https://docs.min.io/docs/minio-server-configuration-guide.html), please confirm?

Federation is a loose concept here @s3rj1k it was only meant to provide a way to manage DNS for your buckets using etcd + coreDNS - if you don't need to do that and your bucket creations are fairly controlled then you can manage this just by using MINIO_DOMAIN env.  and managing bucket DNS  on your DNS provider.

**Comment by harshavardhana** [comment]
@s3rj1k just so you might be interested please look at https://github.com/minio/sidekick

**Comment by s3rj1k** [comment]
@harshavardhana Let me rephrase, by federation I mean a way to horizontally scale cluster size, without dynamic disk management only solution is some kind of sharding with n-clusters.

**Comment by harshavardhana** [comment]
You can do scale by using zones https://github.com/minio/minio/tree/master/docs/distributed#expanding-existing-distributed-setup

**Comment by s3rj1k** [comment]
It is not clear from documentation, adding new zone(cluster) to existing and running instance, would it disrupt operations?

Can I restart distributed cluster nodes one by one without halting cluster operation?

**Comment by harshavardhana** [comment]
You cannot add to a running instance dynamically, you can do rolling add as well of course you will have to change the CLI everywhere.

It is easy to do these things using kubernetes like infrastructure.

**Comment by s3rj1k** [comment]
So I need to stop all instances, change cli params and start all instances for 2 clusters?

**Comment by harshavardhana** [comment]
> So I need to stop all instances, change cli params and start all instances for 2 clusters?

Correct @s3rj1k

**Comment by s3rj1k** [comment]
@harshavardhana Thanks, consider adding this example with stopping all instances to docs.
Closing this issue.
